src/tribler-core/tribler_core/upgrade/tests/test_version_manager.py

```diff
@@ -207,3 +212,98 @@ def test_copy_state_directory(tmpdir):
 
     # Make sure the contents in the before and after upgrade directories are the same
     assert filecmp.cmp(src_dir / 'ec_multichain.pem', tgt_dir / 'ec_multichain.pem')
+
+
+# pylint: disable=too-many-locals
+def test_get_disposable_state_directories(tmpdir_factory):
+    tmpdir = tmpdir_factory.mktemp("scenario")
+    root_state_dir = Path(tmpdir)
+
+    # Scenario: multiple versions of state directory exists, <major.minor.version>. Then on disposable directories
+    # based on the current version show all other directories except the last version.
+
+    major_versions = [8, 7]
+    minor_versions = list(range(10))
+    patch_versions = list(range(3))
+
+    last_version = f"{major_versions[0]}.{minor_versions[-1]}.{patch_versions[0]}"  # 8.9.2
+    last_version_dir = root_state_dir / f"{major_versions[0]}.{minor_versions[-1]}"
+    second_last_version_dir = root_state_dir / f"{major_versions[0]}.{minor_versions[-2]}"
+
+    version_history = {"last_version": last_version, "history": dict()}
+
+    # Create state directories for all older versions
+    base_install_ts = time.time() - 1000  # some timestamp in the past
+    for major in major_versions:
+        for minor in reversed(minor_versions):
+            for patch in patch_versions:
+                version = f"{major}.{minor}.{patch}"
+                version_dir = f"{major}.{minor}"
+
+                # Set install time in order of version. i.e. newer version are installed later
+                version_install_ts = base_install_ts + major * 100 + minor * 10 + patch
+                version_history["history"][version_install_ts] = version
+
+                # Create an empty version directory if does not exist
+                (root_state_dir / version_dir).mkdir(exist_ok=True)
+
+    # Write the version history file before checking disposable directories
+    (root_state_dir / VERSION_HISTORY_FILE).write_text(json.dumps(version_history))
+
+    # Case 0: If the code version and last version is the same, no disposable directories shown at startup
+    code_version = last_version  # 8.9.2
+    disposable_dirs = get_disposable_state_directories(root_state_dir, code_version, skip_last_version=True)
```

Comment:
the output of `get_disposable_state_directories` should be always the same, regardless of whether the code version is there or not. Otherwise, you're mixing _decision_ logic with _analysis_ logic.

Examples: (code version is 8.9.2)
```
8.7
8.8
8.9
disposable_dirs -> [8.7]
```

```
8.7
8.8
disposable_dirs -> []
```

```
8.3
8.7
8.8
disposable_dirs -> []
```